vmdb/app/models/ext_management_system.rb

```diff
@@ -1,20 +1,26 @@
 class ExtManagementSystem < ActiveRecord::Base
+  SUBCLASSES = %w(
+    EmsInfra
+    EmsCloud
+    EmsContainer
+  )
+
   include EmsRefresh::Manager
 
   def self.types
-    [EmsInfra, EmsCloud].collect(&:types).flatten
+    leaf_subclasses.collect(&:ems_type)
   end
 
   def self.supported_types
-    [EmsInfra, EmsCloud].collect(&:supported_types).flatten
+    supported_subclasses.collect(&:ems_type)
   end
 
   def self.leaf_subclasses
-    [EmsInfra, EmsCloud].collect(&:subclasses).flatten
+    descendants.select { |d| d.subclasses.empty? }
   end
 
   def self.supported_subclasses
-    [EmsInfra, EmsCloud].collect(&:supported_subclasses).flatten
+    subclasses.flat_map(&:supported_subclasses)
```

Comment:
This code will all go away eventually once I am done.   This is interim to remove a lot of the existing methods in the subclasses that just aren't needed. So for now,  I just want to leave it. 

The purpose of this code is really questionable anymore anyway.  It used to be used to add EMSes but not have them supported (think kvm).  All of this really just drives the supported_types_and_descriptions method for the UI.

supported_subclasses  is meant to be defined at the middle layer,  so I have to walk subclasses to get it.   For leaves/types,  there is no need to walk,  since the logic to get it is simpler via descendants.